power_grid_model_c/power_grid_model/include/power_grid_model/sparse_ordering.hpp

```diff
@@ -0,0 +1,221 @@
+
+    for (Idx i = 0; i < static_cast<Idx>(l.size()); i++) {
+        IdxVector sl(l.size() - 1);
+        std::copy(l.begin(), l.begin() + i, sl.begin());
+        std::copy(l.begin() + i + 1, l.end(), sl.begin() + i);
+        d[l[i]] = std::move(sl);
+    }
+
+    return d;
+}
+
+inline std::vector<std::pair<IdxVector, IdxVector>> check_indistguishable(Idx const u,
+                                                                          std::map<Idx, IdxVector> const& d) {
+    IdxVector rl;
+
+    auto l = adj(u, d);
+    auto lu = l;
+    lu.push_back(u);
+    std::ranges::sort(lu);
+
+    for (auto const& v : l) {
+        auto lv = adj(v, d);
+        lv.push_back(v);
+        std::ranges::sort(lv);
+        if (lu == lv) {
+            rl.push_back(v);
+        }
+    }
+
+    return {{l, rl}};
+}
+
+inline bool in_graph(std::pair<Idx, Idx> const& e, std::map<Idx, IdxVector> const& d) {
+    if (auto edges_it = d.find(e.first);
+        edges_it != d.cend() && std::ranges::find(edges_it->second, e.second) != edges_it->second.cend()) {
+        return true;
+    }
+    return false;
+}
+
+inline IdxVector remove_vertices_update_degrees(Idx const u, std::map<Idx, IdxVector>& d, DegreeLookup& dgd,
+                                                std::vector<std::pair<Idx, Idx>>& fills) {
+    std::vector<std::pair<IdxVector, IdxVector>> nbsrl = check_indistguishable(u, d);
+    auto& [nbs, rl] = nbsrl[0];
+    IdxVector alpha = rl;
+    std::map<Idx, IdxVector> dd;
+
+    rl.push_back(u);
+
+    for (auto uu : rl) {
+        if (uu != u) {
+            std::erase(nbs, uu);
+        }
+
+        remove_element_degree(uu, dgd);
+        IdxVector el;
+        for (auto& e : d[uu]) {
+            auto& adjacents = d[e];
+            std::erase(adjacents, uu);
+            if (adjacents.empty()) {
+                el.push_back(e);
+            }
+        }
+
+        el.push_back(uu);
+
+        for (auto const& it : el) {
+            d.erase(it);
+        }
+    }
+
+    dd = make_clique(nbs);
+
+    for (auto const& [k, adjacent] : dd) {
+        auto it = d.find(k);
+        for (Idx const e : adjacent) {
+            if (!in_graph(std::make_pair(k, e), d)) {
+                if (it == d.end()) {
+                    std::tie(it, std::ignore) = d.try_emplace(k);
+                }
+                it->second.push_back(e);
+                d[e].push_back(k);
+                fills.emplace_back(k, e);
+            }
+        }
+    }
+
+    for (auto const& e : nbs) {
+        set_element_degree(e, num_adjacent(e, d), dgd);
+    }
+
+    return alpha;
+}
+} // namespace detail
+
+inline std::pair<IdxVector, std::vector<std::pair<Idx, Idx>>> minimum_degree_ordering(std::map<Idx, IdxVector> d_) {
+    // make symmetric
+    auto d = d_;
+    for (auto& [k, adjacent] : d_) {
+        for (auto a : adjacent) {
+            d[a].push_back(k);
+        }
+    }
+    for (auto& [k, adjacent] : d) {
+        std::ranges::sort(adjacent);
+    }
```

Comment:
i can also do this in the data prep cause this is useless overhead, but it will change the input and therefore the tests

Reply:
Nice improvement so far! I think we can merge this PR once ci is fixed

Reply:
i decided to remove the copy here, and instead go with dedup using `set` to fix this. that basically removes the precondition that this piece of code introduced without a large weird copy on the input